### PR TITLE
Change Array::MakeRef to use MakeAlias

### PR DESCRIPTION
**Pull request by pazner**
- Change `Array::MakeRef` to use `MakeAlias`
- Allow multiple test cases in debug device tests (hopefully did this right with `makefile` and `CMakeLists.txt`)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3579](https://github.com/mfem/mfem/pull/3579) | @pazner | @v-dobrev | @v-dobrev + @camierjs | 4/6/23 | 4/11/23 | 4/15/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: tests/unit/miniapps/test_debug_device.cpp [hunk 1]
```diff
@@ -87,6 +88,15 @@ void Aliases(const int N = 0x1234)
    REQUIRE(S*S == MFEM_Approx(24.0*N));
 }
 
+TEST_CASE("Array::MakeRef", "[DebugDevice]")
+{
+   Array<int> x(1);
+   Array<int> y;
+   y.MakeRef(x);
+   x.Read();
+   REQUIRE_NOTHROW(y.Read());
```
Comment (v-dobrev) [inline tests/unit/miniapps/test_debug_device.cpp:97]: I'm not sure we want to enforce this: after `x.Read()`, `y` is out of sync, so requiring that `y.Read()` works without a sync is a strong requirement.

@camierjs, what do you think?

Comment (camierjs) [reply tests/unit/miniapps/test_debug_device.cpp:97]: I'm fine leaving it there as it is not throwing the memory error.

Comment (v-dobrev) [reply tests/unit/miniapps/test_debug_device.cpp:97]: Okay, let's keep it as it is.

**Comment by v-dobrev** [comment]
I'm getting some failures from `make test` when using a debug build:
```console
...
    Parallel unit tests [ mpirun -np 1 psedov_tests_debug ... ]: FAILED  (2.81s 17264kB)
INFO: Test filter: [Parallel] ~[MFEMData] 
Filters: [Parallel] ~[MFEMData]
Zones min/max: 4 4
Number of local/global kinematic (position, velocity) dofs: 50/50
Number of local/global specific internal energy dofs: 16/16
step     5,	t = 0.2026,	dt = 0.040687,	|e| = 3.5082549452
step    10,	t = 0.4061,	dt = 0.040687,	|e| = 3.0482318525
step    15,	t = 0.6000,	dt = 0.031804,	|e| = 2.7564445968


Assertion failed: (registered || !(owns_host || owns_device || owns_internal) || (!(owns_device || owns_internal) && h_ptr == nullptr)) is false:
 --> invalid Memory state
 ... in function: static void mfem::MemoryManager::Delete_(void *, mfem::MemoryType, unsigned int)
 ... in file: general/mem_manager.cpp:927

Backtrace:
0) [0x473e000102fbfa1b]: mfem::mfem_error(char const*)
1) [0x335e000102fcc80f]: mfem::MemoryManager::Delete_(void*, mfem::MemoryType, unsigned int)
2) [0xa25300010303c8c7]: mfem::ConstrainedOperator::~ConstrainedOperator()
3) [0xdb3f80010303c483]: mfem::ConstrainedOperator::~ConstrainedOperator()
4) [0x4756800102f9b0cb]: mfem::hydrodynamics::PAMassOperator::~PAMassOperator()
5) [0x144d000102f9e343]: mfem::hydrodynamics::LagrangianHydroOperator::~LagrangianHydroOperator()
6) [0x9978000102f8bee3]: mfem::sedov(int, int, char**)
7) [0x4959800102f8cd3b]: C_A_T_C_H_T_E_S_T_0()
8) [0x840b000102f4a15f]: Catch::TestInvokerAsFunction::invoke() const
9) [0xe1d800102f42dfb]: Catch::RunContext::invokeActiveTestCase()
10) [0xc751000102f4109b]: Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&)
11) [0x494a000102f4078b]: Catch::RunContext::runTest(Catch::TestCase const&)
12) [0x886a800102f4602b]: Catch::Session::runInternal()
13) [0x9562800102f4552f]: Catch::Session::run()
14) [0xaf66000102f5d16b]: main

Lookup backtrace source lines:
atos -o /<truncated>/mfem/tests/unit/psedov_tests_debug -l 0x102f28000 0x473e000102fbfa1b 0x335e000102fcc80f 0xa25300010303c8c7 0xdb3f80010303c483 0x4756800102f9b0cb 0x144d000102f9e343 0x9978000102f8bee3 0x4959800102f8cd3b 0x840b000102f4a15f 0xe1d800102f42dfb 0xc751000102f4109b 0x494a000102f4078b 0x886a800102f4602b 0x9562800102f4552f 0xaf66000102f5d16b

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
make[1]: *** [psedov_tests_debug-test-par] Error 1
...
    Unit tests [ sedov_tests_debug ... ]: FAILED  (0.72s 10704kB)
INFO: Test filter: ~[Parallel] ~[MFEMData] 
Filters: ~[Parallel] ~[MFEMData]
Zones min/max: 4 4
Number of local/global kinematic (position, velocity) dofs: 50/50
Number of local/global specific internal energy dofs: 16/16
step     5,	t = 0.2026,	dt = 0.040687,	|e| = 3.5082549452
step    10,	t = 0.4061,	dt = 0.040687,	|e| = 3.0482318525
step    15,	t = 0.6000,	dt = 0.031804,	|e| = 2.7564445968


Assertion failed: (registered || !(owns_host || owns_device || owns_internal) || (!(owns_device || owns_internal) && h_ptr == nullptr)) is false:
 --> invalid Memory state
 ... in function: static void mfem::MemoryManager::Delete_(void *, mfem::MemoryType, unsigned int)
 ... in file: general/mem_manager.cpp:927

Backtrace:
0) [0xab54000102a37d5f]: mfem::mfem_error(char const*)
1) [0xe72d000102a44b53]: mfem::MemoryManager::Delete_(void*, mfem::MemoryType, unsigned int)
2) [0x664a000102ab4cef]: mfem::ConstrainedOperator::~ConstrainedOperator()
3) [0xb438000102ab486b]: mfem::ConstrainedOperator::~ConstrainedOperator()
4) [0x8260000102a13463]: mfem::hydrodynamics::PAMassOperator::~PAMassOperator()
5) [0xe71c800102a164ff]: mfem::hydrodynamics::LagrangianHydroOperator::~LagrangianHydroOperator()
6) [0xc47b000102a04147]: mfem::sedov(int, int, char**)
7) [0x9c7d000102a04f0f]: C_A_T_C_H_T_E_S_T_0()
8) [0x68148001029c2bdf]: Catch::TestInvokerAsFunction::invoke() const
9) [0x90498001029bba2b]: Catch::RunContext::invokeActiveTestCase()
10) [0xf91a0001029b9d0f]: Catch::RunContext::runCurrentTest(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&)
11) [0x35768001029b93ff]: Catch::RunContext::runTest(Catch::TestCase const&)
12) [0xa70d0001029beab7]: Catch::Session::runInternal()
13) [0x675e8001029bdff7]: Catch::Session::run()
14) [0xd9388001029d5b1f]: main

Lookup backtrace source lines:
atos -o /<truncated>/mfem/tests/unit/sedov_tests_debug -l 0x1029a0000 0xab54000102a37d5f 0xe72d000102a44b53 0x664a000102ab4cef 0xb438000102ab486b 0x8260000102a13463 0xe71c800102a164ff 0xc47b000102a04147 0x9c7d000102a04f0f 0x68148001029c2bdf 0x90498001029bba2b 0xf91a0001029b9d0f 0x35768001029b93ff 0xa70d0001029beab7 0x675e8001029bdff7 0xd9388001029d5b1f


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sedov_tests_debug is a Catch v2.13.9 host application.
Run with -? for options

-------------------------------------------------------------------------------
Sedov
-------------------------------------------------------------------------------
/<truncated>/mfem/tests/unit/miniapps/test_sedov.cpp:2212
...............................................................................

/<truncated>/mfem/tests/unit/miniapps/test_sedov.cpp:2212: FAILED:
  {Unknown expression after the reported line}
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

===============================================================================
test cases:   1 |   0 passed | 1 failed
assertions: 140 | 139 passed | 1 failed

make[1]: *** [sedov_tests_debug-test-seq] Error 134
...
```
Maybe the `MFEM_ASSERT` is not correct?

**Comment by v-dobrev** [comment]
I think I fixed the above bug -- registering an alias to `nullptr` (e.g. an alias to an empty `Array`) was creating a `Memory` where the `h_ptr` is `nullptr` and the `OWNS_INTERNAL` flag was true which is incorrect since we do not actually register `nullptr` in the map of all aliases.

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing.

**Pull request merged**